### PR TITLE
761 better getter seeters@221 fix separator div@main

### LINKED ISSUE #761: Add getter/setter for section dividers
Create new getter (`section_div()`) and setter (`section_div<-()`) for section dividers.

### PR DESCRIPTION
**Pull request by Melkiades**
Fixes #761 #762. Somehow close to #509 

```r
basic_table(header_section_div = " ") %>%
    split_cols_by("ARM") %>%
    split_rows_by("SEX", split_fun = drop_split_levels) %>%
    analyze("AGE") %>%
    build_table(DM)
```
makes
```
         A: Drug X   B: Placebo   C: Combination
————————————————————————————————————————————————
                                                
F                                               
  Mean     33.71       33.84          34.89     
M                                               
  Mean     36.55       32.10          34.28     
```
but if there are multiple analyze it does not print anything. These options differ, but it should print something in those cases.

### CODE REVIEW

**Review by edelarua — COMMENTED** [review]
Great job! I think this would cover a wide variety of use cases if people want to use section dividers, but is there any way to produce this table:

```
                 all obs
————————————————————————
A: Drug X               
  A                     
    Mean          32.53 
  B                     
    Mean          35.46 
  C                     
    Mean          36.34 
------------------------
B: Placebo              
  A                     
    Mean          32.30 
  B                     
    Mean          32.42 
  C                     
    Mean          34.45 
------------------------
C: Combination          
  A                     
    Mean          35.76 
  B                     
    Mean          34.39 
  C                     
    Mean          33.54 
```

I feel like this would be the most typical implementation by basic users.

File: R/00tabletrees.R [hunk 1]
```diff
@@ -647,7 +646,8 @@ AnalyzeVarSplit <- function(var,
                             extra_args = list(),
                             indent_mod = 0L,
                             label_pos = "default",
-                            cvar = "") {
+                            cvar = "",
+                            section_div = NA_character_) {
```
Comment (Melkiades) [inline R/00tabletrees.R:650]: Now it is possible to do it also in analyze

File: R/tt_paginate.R [hunk 2]
```diff
@@ -257,8 +254,10 @@ setMethod(
 
 
     if (NROW(content_table(tt)) > 0) {
-      cind <- indent + indent_mod(content_table(tt))
-      contdf <- make_row_df(content_table(tt),
+      ct_tt <- content_table(tt)
+      cind <- indent + indent_mod(ct_tt)
+      trailing_section_div(ct_tt) <- trailing_section_div(tt_labelrow(tt))
```
Comment (Melkiades) [inline R/tt_paginate.R:259]: As section_div setter is based solely on labelrow and datarow, this fixes the contentRow case

File: tests/testthat/test-accessors.R [hunk 3]
```diff
@@ -223,3 +223,140 @@ test_that("header sep setting works", {
...
+  expect_true(check_pattern(header_sdiv, "+", nchar(header_sdiv)))
+})
+
+check_pattern <- function(element, letter, len) {
+  # Regular expression to match exactly three of the same letter
+  regex <- paste0("^", letter, "{", len,"}$")
+  return(grepl(regex, element))
+}
+
+test_structure_with_a_getter <- function(tbl, getter, val_per_lev) {
```
Comment (Melkiades) [inline tests/testthat/test-accessors.R:334]: helper function to check we are doing the right thing

Comment (Melkiades) [reply tests/testthat/test-accessors.R:334]: to check vs structure

**Comment by Melkiades** [comment]
> Great job! I think this would cover a wide variety of use cases if people want to use section dividers, but is there any way to produce this table:
> 
> ```
>                  all obs
> ————————————————————————
> A: Drug X               
>   A                     
>     Mean          32.53 
>   B                     
>     Mean          35.46 
>   C                     
>     Mean          36.34 
> ------------------------
> B: Placebo              
>   A                     
>     Mean          32.30 
>   B                     
>     Mean          32.42 
>   C                     
>     Mean          34.45 
> ------------------------
> C: Combination          
>   A                     
>     Mean          35.76 
>   B                     
>     Mean          34.39 
>   C                     
>     Mean          33.54 
> ```
> 
> I feel like this would be the most typical implementation by basic users.

```r
basic_table() %>% 
  split_rows_by("ARM", section_div = "-") %>% 
  split_rows_by("STRATA1") %>% 
  analyze("BMRKR1") %>% 
  build_table(DM)
```

Can you please see if section_div getter and setter work fine for you? :)

**Comment by edelarua** [comment]
Hi Davide,

Thanks for the example code - I think my test code was getting too complicated and I missed that somehow. 

A couple of questions/comments on the getters/setters:

I'll use this table code for examples:
```r
t <- basic_table() %>% 
    split_rows_by("ARM") %>% 
    split_rows_by("STRATA1") %>% 
    analyze("BMRKR1") %>% 
    build_table(DM)
```
```
                 all obs
————————————————————————
A: Drug X               
  A                     
    Mean          5.90  
  B                     
    Mean          5.55  
  C                     
    Mean          5.94  
B: Placebo              
  A                     
    Mean          6.01  
  B                     
    Mean          5.86  
  C                     
    Mean          6.53  
C: Combination          
  A                     
    Mean          5.95  
  B                     
    Mean          6.32  
  C                     
    Mean          4.91  
```

1. I'm not sure if this is a lot more difficult to do (in which case you can ignore me :) ) but I think the most logical way to implement `section_div` would be to go from the outermost to innermost split dividers. So if the user gave only one value in the example table `t` above (i.e. `section_div(t) <- "~"`), the section divider would be applied between arm splits, if they gave two (i.e. `section_div(t) <- c("~", ".")`) the first would apply to the arm split, the second would apply to the STRATA1 split, etc.
In the current implementation, you have it requiring a vector of length `nrow(t)` which I feel might be a bit overwhelming to the average user. But again, it's working great as is for me, this is just an alternative idea - I'm not sure how specifically a typical user would want to edit the section dividers, or if they need them in weird places :).

2. Not sure if you meant for me to test `header_section_div` but it's currently adding a line _below_ what I would expect to be the header section div (i.e. the line separating the header contents from the body contents).

E.g.
```r
header_section_div(t) <- "~"
t
```
```
                 all obs
————————————————————————
~~~~~~~~~~~~~~~~~~~~~~~~
A: Drug X               
  A                     
    Mean          5.90  
  B                     
    Mean          5.55  
  C                     
    Mean          5.94  
B: Placebo              
  A                     
    Mean          6.01  
  B                     
    Mean          5.86  
  C                     
    Mean          6.53  
C: Combination          
  A                     
    Mean          5.95  
  B                     
    Mean          6.32  
  C                     
    Mean          4.91  
```

Other than that the getters/setters are all working correctly for me!

**Comment by Melkiades** [comment]
Thanks for trying it out Emily! 
1. It makes sense, but I needed to get the row-wise implementation first to establish exactly where everything should be to set/get the whole tree divisors. It will take a bit of work, but I can add separators for splits only and one value only for those splits. 
2. I think we need another variable for #509 as the one I implemented was asked by Chenkai and others some times ago and it was to add a space between header line and table body

**Comment by Melkiades** [comment]
@edelarua 1. Is done. I added as default that if you add only one value, you have it for all levels. Let me know if only the first split/section should have it. You can do that anyway like this: `section_div(tbl) <- c("a", NA_character_)`. 

2. as mentioned, `header_section_div` is another thing, i.e. the need to have a `" "` before the start of the table. There is no other way to add this empty line otherwise. For the table divisor, instead, we can still use the `section_div` and put it where we want. Doing it with another section div parameter seems a bit difficult, also considering the fact that the print collapses it and the realization could be redundant (and you need to understand much better how tables works to attempt this thing specifically)

File: R/tree_accessors.R [hunk 4]
```diff
@@ -3195,15 +3185,292 @@ setMethod(
...
+
+#' @title Section dividers setter ang getter
+#'
+#' @description
+#' `section_div`can be used to set or get the section divider for a table object
+#' produced by [build_table()]. When assigned in post processing (`section_div<-`)
+#' the table can have a section divider after each row, independently. Otherwise,
+#' only [split_rows_by()] (and its related row-wise splits) and [analyze()] have a
+#' `section_div` parameters that will produce separators between split sections, or
+#' data subgroups.
```
Comment (edelarua) [inline R/tree_accessors.R:3216]:
```suggestion
#' `section_div` can be used to set or get the section divider for a table object
#' produced by [build_table()]. When assigned in post-processing (`section_div<-`)
#' the table can have a section divider after every row, each assigned independently.
#' If assigning during layout creation, only [split_rows_by()] (and its related row-wise
#' splits) and [analyze()] have a `section_div` parameter that will produce separators 
#' between split sections and data subgroups, respectively.
```

File: R/tree_accessors.R [hunk 5]
```diff
@@ -3195,15 +3185,292 @@ setMethod(
...
+#' @seealso [basic_table()] parameter `header_section_div` for a global section divider.
+#'
+#' @details
+#' Assigned value to section divider must be a character vector. If any value is `NA_character_`
+#' the section divider will be absent for that row or section. When you want to only affect sections
+#' or splits, please use `only_sep_sections` or provide a shorter vector than the number of rows. 
+#' Ideally, the length of the vector should be less than the number of splits with, eventually, the
+#' leaf-level, i.e. `DataRow` where analyze results are. Note that if only one value is inserted,
+#' only the first split will be affected.
+#' If `only_sep_sections =TRUE`, which is the default for `section_div()` produced from the table
```
Comment (edelarua) [inline R/tree_accessors.R:3241]:
```suggestion
#' If `only_sep_sections = TRUE`, which is the default for `section_div()` produced from the table 
```

File: R/tree_accessors.R [hunk 6]
```diff
@@ -3195,15 +3185,292 @@ setMethod(
...
+setMethod("trailing_section_div<-", "LabelRow", function(obj, value) {
+  obj@trailing_section_div <- value
+  obj
+})
+setMethod("trailing_section_div<-", "TableRow", function(obj, value) {
+  obj@trailing_section_div <- value
+  obj
+})
+
+#' @title Section dividers setter ang getter
```
Comment (edelarua) [inline R/tree_accessors.R:3208]:
```suggestion
#' @title Section dividers getter and setter
```

File: R/tree_accessors.R [hunk 7]
```diff
@@ -3195,15 +3185,292 @@ setMethod(
...
+#' only [split_rows_by()] (and its related row-wise splits) and [analyze()] have a
+#' `section_div` parameters that will produce separators between split sections, or
+#' data subgroups.
+#'
+#' @param obj Table object. This can be of any class that inherits from `VTableTree`
+#'   or `TableRow`/`LabelRow`.
+#' @param only_sep_sections logical(1). Defaults to `FALSE` for `section_div<-`. It allows
+#'   to set the section divider only for sections that are splits or analyses if the number of
+#'   values is less than the number of rows in the table. If `TRUE`, the section divider will
+#'   be set for all the rows of the table.
```
Comment (edelarua) [inline R/tree_accessors.R:3223]:
```suggestion
#' @param only_sep_sections logical(1). Defaults to `FALSE` for `section_div<-`. Allows
#'   you to set the section divider only for sections that are splits or analyses if the number of
#'   values is less than the number of rows in the table. If `TRUE`, the section divider will
#'   be set for all rows of the table.
```

File: R/tree_accessors.R [hunk 8]
```diff
@@ -3195,15 +3185,292 @@ setMethod(
...
+#' @param value character. The section divider single character vector. Each character that is
+#'   not `NA_character_` will produce a trailing separator for each row of the table. `value` length
+#'   should reflect the number of rows, or be between 1 and the number of splits/levels. 
+#'   See details for more information.
```
Comment (edelarua) [inline R/tree_accessors.R:3227]:
```suggestion
#' @param value character. Vector of single characters to use as section dividers. Each character 
#'   is repeated such that all section dividers span the width of the table. Each character that is
#'   not `NA_character_` will produce a trailing separator for each row of the table. `value` length
#'   should reflect the number of rows, or be between 1 and the number of splits/levels. 
#'   See the Details section below for more information.
```

File: NEWS.md [hunk 9]
```diff
@@ -1,13 +1,22 @@
...
+   related`basic_table` parameter.
  
+### Bug Fixes
+ * Fixed a bug that was failing when wrapping and section dividers were used at the same time.
 
 ### Miscellaneous
  * Applied `styler` and resolved package lint. Changed default indentation from 4 spaces to 2.
  * Added Developer Guide to pkgdown site with Debugging, Split Machinery, and Tabulation sections.
  * Whitespace is not trimmed when rendering tables with `as_html`.
+ * Exported `section_div` methods have now a dedicated documentation page that is visible to users.
```
Comment (edelarua) [inline NEWS.md:19]:
```suggestion
 * Exported `section_div` methods now have a dedicated documentation page that is visible to users.
```

File: NEWS.md [hunk 10]
```diff
@@ -1,13 +1,22 @@
 ## rtables 0.6.5.9011
 ### New Features
  * Removed `ref_group` reordering in column splits so not to change the order.
- * Added `bold` argument to `as_html` to bold specified elements, and `header_sep_line` argument to print a horizontal line under the table header in rendered HTML output.
+ * Added `bold` argument to `as_html` to bold specified elements, and `header_sep_line` 
+   argument to print a    horizontal line under the table header in rendered HTML output.
```
Comment (edelarua) [inline NEWS.md:5]:
```suggestion
 * Added `bold` argument to `as_html` to bold specified elements, and `header_sep_line` 
   argument to print a horizontal line under the table header in rendered HTML output.
```

File: NEWS.md [hunk 11]
```diff
@@ -1,13 +1,22 @@
...
+ * Section divisors can be set also for analysis rows.
+ * Added setter and getter for section dividers (`section_div` and `section_div<-`). They accept also
+   split section structure assignment.
+ * Added `header_section_div` setters and getters for layout and table objects along with 
+   related`basic_table` parameter.
```
Comment (edelarua) [inline NEWS.md:10]:
```suggestion
   related `basic_table` parameter.
```

File: NEWS.md [hunk 12]
```diff
@@ -1,13 +1,22 @@
 ## rtables 0.6.5.9011
 ### New Features
  * Removed `ref_group` reordering in column splits so not to change the order.
- * Added `bold` argument to `as_html` to bold specified elements, and `header_sep_line` argument to print a horizontal line under the table header in rendered HTML output.
+ * Added `bold` argument to `as_html` to bold specified elements, and `header_sep_line` 
+   argument to print a    horizontal line under the table header in rendered HTML output.
+ * Section divisors can be set also for analysis rows.
```
Comment (edelarua) [inline NEWS.md:6]:
```suggestion
 * Section dividers can be set for analysis rows.
```

File: NEWS.md [hunk 13]
```diff
@@ -1,13 +1,22 @@
...
+ * Added setter and getter for section dividers (`section_div` and `section_div<-`). They accept also
```
Comment (edelarua) [inline NEWS.md:7]:
```suggestion
 * Added setter and getter for section dividers (`section_div` and `section_div<-`). They also accept
```

**Review by edelarua — APPROVED** [review]
The new behaviour is great, thanks for implementing that!! Everything seems to be working as expected for me, and the warning messages are very informative. I reviewed the documentation as well and just added some minor suggestions for grammar fixes/typos.

Excellent work!!! :)

File: NEWS.md (hunk 9 above)
Comment (edelarua) [reply NEWS.md]: Updated NEWS when merging in `main`

**Pull request merged**